### PR TITLE
In-memory Engine: fix panic when pending region split before start batch loading

### LINKED ISSUE #17403: In-memory engine: pending region split may cause tikv panic
## Bug Report

https://github.com/tikv/tikv/blob/9734214510548db97c6ff61c3af52fb9dee13de0/components/range_cache_memory_engine/src/background.rs#L925-L940

The current implementation will update region's state from `ReadyToLoad` to `Loading` before start batch loading. But target region may be split after the task is scheduled by before loading start. So this update only update the region with the same id with source region, but not other split region in the same range.

This will cause panic in `on_snapshot_load_finished` as it assumes all the regions in the range should be in `Loading` state.

### PR DESCRIPTION
**Pull request by glorv**
### What is changed and how it works?

Issue Number: ref #16141, close #17403

What's Changed:

```commit-message
Fix the bug that pending region split before batch loading start can
cause panic.

When handle LoadRegion event, the background worker only update the
region with the same id to Loading state, but the region is possible
already split so other region within the same range will still in
`ReadyToLoad` state. So it will cause panic in
`on_snapshot_load_finished`.

The PR removes the useless intermediate state `ReadyToLoad`. After this
change, the raft apply worker will directly update state from Pending
to Loading before scheduler LoadRegion task. So the background worker
will not try to update target region's state.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/range_cache_memory_engine/tests/failpoints/test_memory_engine.rs [hunk 1]
```diff
@@ -645,3 +645,101 @@ fn test_load_with_gc() {
         .unwrap_err();
     engine.snapshot(region.id, 0, range, 7, 100).unwrap();
 }
+
+#[test]
+fn test_region_split_before_batch_loading_start() {
```
Comment (overvenus) [inline components/range_cache_memory_engine/tests/failpoints/test_memory_engine.rs:650]: Please add some comment about this test.

**Review by overvenus — COMMENTED** [review]
Rest LGTM

**Comment by overvenus** [comment]
By the way, please file a bug issue.

File: components/range_cache_memory_engine/tests/failpoints/test_memory_engine.rs (hunk 1 above)
Comment (glorv) [reply components/range_cache_memory_engine/tests/failpoints/test_memory_engine.rs:654]: Done

**Review by overvenus — APPROVED** [review]
LGTM

**Pull request merged**